lib/ex_unit/lib/ex_unit/capture_server.ex

```diff
@@ -83,25 +84,71 @@ defmodule ExUnit.CaptureServer do
     {:noreply, config}
   end
 
-  defp release_device(ref, %{devices: {names, refs}} = config) do
-    case Map.pop(refs, ref) do
-      {{name, pid}, refs} ->
-        names = Map.delete(names, name)
-        Process.demonitor(ref, [:flush])
-
-        try do
-          try do
-            Process.unregister(name)
-          after
-            Process.register(pid, name)
-          end
-        rescue
-          ArgumentError -> nil
-        end
+  defp capture_device(name, encoding, "", %{devices: devices} = config, caller)
+       when :erlang.is_map_key(name, devices) do
+    case Map.get(devices, name) do
+      %{encoding: ^encoding} = device ->
+        {_, output} = StringIO.contents(device.pid)
+        ref = Process.monitor(caller)
+        refs = [{ref, byte_size(output)} | device.refs]
+        devices = %{devices | name => %{device | refs: refs}}
+        {:reply, {:ok, ref}, %{config | devices: devices}}
+
+      %{encoding: other_encoding} ->
+        {:reply, {:error, {:changed_encoding, other_encoding}}, config}
+    end
+  end
+
+  defp capture_device(name, _, _, %{devices: devices} = config, _)
+       when :erlang.is_map_key(name, devices) do
+    {:reply, {:error, :input_on_already_captured_device}, config}
+  end
+
+  defp capture_device(name, encoding, input, %{devices: devices} = config, caller) do
+    {:ok, pid} = StringIO.open(input, encoding: encoding)
+    orig_pid = Process.whereis(name)
 
-        %{config | devices: {names, refs}}
+    Process.unregister(name)
+    Process.register(pid, name)
 
-      {nil, _refs} ->
+    _ = Process.monitor(pid)
+    ref = Process.monitor(caller)
+
+    device = %{orig_pid: orig_pid, pid: pid, refs: [{ref, 0}], encoding: encoding}
```

Comment:
No abbreviations please. :)

Reply:
`refs: [{ref, 0}]` should be a map too, it will make it easier to read and remove refs.